src/corecel/data/detail/PinnedAllocator.hh

```diff
@@ -0,0 +1,49 @@
+//----------------------------------*-C++-*----------------------------------//
+// Copyright 2023 UT-Battelle, LLC, and other Celeritas developers.
+// See the top-level COPYRIGHT file for details.
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file corecel/data/detail/PinnedAllocator.hh
+//---------------------------------------------------------------------------//
+#pragma once
+
+#include <cstdlib>
+#include <limits>
+#include <new>
+
+#include "corecel/Assert.hh"
+#include "corecel/Macros.hh"
+#include "corecel/device_runtime_api.h"
+
+namespace celeritas
+{
+namespace detail
+{
+template<class T>
+struct PinnedAllocator
```

Comment:
@sethrj Pimpl vs explicit instantiation to hide the CUDA/HIP dependency? I think I like the latter more, the only downside is that we'd have to specify instantiation for quite a few types.